In [2]:
import pandas as pd 
import numpy as np

In [3]:
df = pd.read_csv("D:\\João Carlos Moura\\brasileirao-data-analysis\\data\\raw\\brasileirao.csv")

df.head()

,ano_campeonato,data,rodada,estadio,arbitro,publico,publico_max,time_mandante,time_visitante,tecnico_mandante,...,chutes_bola_parada_mandante,chutes_bola_parada_visitante,defesas_mandante,defesas_visitante,impedimentos_mandante,impedimentos_visitante,chutes_mandante,chutes_visitante,chutes_fora_mandante,chutes_fora_visitante
0,2005,2005-10-08,31,Estádio Durival Britto e Silva,NaN,NaN,NaN,Paraná,Fluminense,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2005,2005-04-23,1,Estádio Durival Britto e Silva,NaN,NaN,NaN,Paraná,Goiás EC,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2005,2005-08-21,21,Estádio Durival Britto e Silva,NaN,NaN,NaN,Paraná,Vasco da Gama,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2005,2005-08-28,23,Estádio Durival Britto e Silva,NaN,NaN,NaN,Paraná,São Paulo,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2005,2005-06-26,9,Estádio Olímpico Nilton Santos,NaN,NaN,NaN,Botafogo,Figueirense FC,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [4]:
df.shape

(8453, 35)

In [5]:
df.columns

Index(['ano_campeonato', 'data', 'rodada', 'estadio', 'arbitro', 'publico',
       'publico_max', 'time_mandante', 'time_visitante', 'tecnico_mandante',
       'tecnico_visitante', 'colocacao_mandante', 'colocacao_visitante',
       'valor_equipe_titular_mandante', 'valor_equipe_titular_visitante',
       'idade_media_titular_mandante', 'idade_media_titular_visitante',
       'gols_mandante', 'gols_visitante', 'gols_1_tempo_mandante',
       'gols_1_tempo_visitante', 'escanteios_mandante', 'escanteios_visitante',
       'faltas_mandante', 'faltas_visitante', 'chutes_bola_parada_mandante',
       'chutes_bola_parada_visitante', 'defesas_mandante', 'defesas_visitante',
       'impedimentos_mandante', 'impedimentos_visitante', 'chutes_mandante',
       'chutes_visitante', 'chutes_fora_mandante', 'chutes_fora_visitante'],
      dtype='str')

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8453 entries, 0 to 8452
Data columns (total 35 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ano_campeonato                  8453 non-null   int64  
 1   data                            8453 non-null   str    
 2   rodada                          8453 non-null   int64  
 3   estadio                         8442 non-null   str    
 4   arbitro                         6743 non-null   str    
 5   publico                         6742 non-null   float64
 6   publico_max                     4181 non-null   float64
 7   time_mandante                   8453 non-null   str    
 8   time_visitante                  8453 non-null   str    
 9   tecnico_mandante                6299 non-null   str    
 10  tecnico_visitante               6299 non-null   str    
 11  colocacao_mandante              6743 non-null   float64
 12  colocacao_visitante             6743 non-null

In [7]:
df.describe


<bound method NDFrame.describe of       ano_campeonato        data  rodada  \
0               2005  2005-10-08      31   
1               2005  2005-04-23       1   
2               2005  2005-08-21      21   
3               2005  2005-08-28      23   
4               2005  2005-06-26       9   
...              ...         ...     ...   
8448            2017  2017-10-15      28   
8449            2017  2017-10-22      30   
8450            2017  2017-10-28      31   
8451            2017  2017-11-09      33   
8452            2017  2017-11-19      36   

                                          estadio                   arbitro  \
0                  Estádio Durival Britto e Silva                       NaN   
1                  Estádio Durival Britto e Silva                       NaN   
2                  Estádio Durival Britto e Silva                       NaN   
3                  Estádio Durival Britto e Silva                       NaN   
4                  Estádio Olímpico Nilton

In [8]:
df.isnull().sum()

ano_campeonato                       0
data                                 0
rodada                               0
estadio                             11
arbitro                           1710
publico                           1711
publico_max                       4272
time_mandante                        0
time_visitante                       0
tecnico_mandante                  2154
tecnico_visitante                 2154
colocacao_mandante                1710
colocacao_visitante               1710
valor_equipe_titular_mandante     2098
valor_equipe_titular_visitante    2098
idade_media_titular_mandante      2100
idade_media_titular_visitante     2100
gols_mandante                        1
gols_visitante                       1
gols_1_tempo_mandante             1721
gols_1_tempo_visitante            1721
escanteios_mandante               6366
escanteios_visitante              6366
faltas_mandante                   6366
faltas_visitante                  6366
chutes_bola_parada_mandan

In [9]:
df['ano_campeonato']

0       2005
1       2005
2       2005
3       2005
4       2005
        ... 
8448    2017
8449    2017
8450    2017
8451    2017
8452    2017
Name: ano_campeonato, Length: 8453, dtype: int64

In [10]:
df['arbitro']

0                            NaN
1                            NaN
2                            NaN
3                            NaN
4                            NaN
                  ...           
8448    Marcelo de Lima Henrique
8449                Rafael Traci
8450            Anderson Daronco
8451    Marcelo de Lima Henrique
8452              Wilton Sampaio
Name: arbitro, Length: 8453, dtype: str

In [11]:
df[['ano_campeonato', 'arbitro']]

,ano_campeonato,arbitro
0,2005,NaN
1,2005,NaN
2,2005,NaN
3,2005,NaN
4,2005,NaN
...,...,...
8448,2017,Marcelo de Lima Henrique
8449,2017,Rafael Traci
8450,2017,Anderson Daronco
8451,2017,Marcelo de Lima Henrique


## Observações iniciais

- A base contém **8453 partidas** e **35 colunas**.
- Os dados abrangem diversas temporadas do Campeonato Brasileiro.
- Algumas colunas apresentam valores ausentes, principalmente **arbitro**, indicando informações não registradas em determinadas partidas.
- Não foram encontradas linhas duplicadas.
- A coluna **data** foi convertida para o formato datetime.
- As estatísticas indicam que os gols e o público possuem valores coerentes, sem números negativos.
- A base contém informações detalhadas, incluindo estatísticas ofensivas, defensivas, escanteios, faltas, chutes e dados dos elencos.
- A riqueza das variáveis permitirá realizar análises descritivas, comparações entre equipes e, futuramente, modelos preditivos.

In [12]:
df_clean = df.copy()

In [13]:
df_clean["data"] = pd.to_datetime(df_clean["data"])

In [14]:
df_clean["data"].dtype


dtype('<M8[us]')

In [15]:
df_clean.duplicated().sum()

np.int64(0)

#Não foram identificadas linhas duplicadas na base de dados


In [16]:
missing = df_clean.isnull().sum()

missing[missing > 0].sort_values(ascending=False)

chutes_fora_visitante             6366
chutes_fora_mandante              6366
impedimentos_visitante            6366
impedimentos_mandante             6366
chutes_visitante                  6366
chutes_mandante                   6366
defesas_mandante                  6366
defesas_visitante                 6366
chutes_bola_parada_visitante      6366
chutes_bola_parada_mandante       6366
faltas_mandante                   6366
faltas_visitante                  6366
escanteios_mandante               6366
escanteios_visitante              6366
publico_max                       4272
tecnico_mandante                  2154
tecnico_visitante                 2154
idade_media_titular_visitante     2100
idade_media_titular_mandante      2100
valor_equipe_titular_visitante    2098
valor_equipe_titular_mandante     2098
gols_1_tempo_visitante            1721
gols_1_tempo_mandante             1721
publico                           1711
arbitro                           1710
colocacao_mandante       

In [17]:
missing_percent = (
    df_clean.isnull().sum() / len(df_clean) * 100
).sort_values(ascending=False)

missing_percent[missing_percent > 0]

chutes_fora_visitante             75.310541
defesas_visitante                 75.310541
defesas_mandante                  75.310541
faltas_mandante                   75.310541
faltas_visitante                  75.310541
chutes_bola_parada_visitante      75.310541
chutes_fora_mandante              75.310541
impedimentos_visitante            75.310541
impedimentos_mandante             75.310541
chutes_visitante                  75.310541
chutes_mandante                   75.310541
escanteios_mandante               75.310541
escanteios_visitante              75.310541
chutes_bola_parada_mandante       75.310541
publico_max                       50.538270
tecnico_mandante                  25.482077
tecnico_visitante                 25.482077
idade_media_titular_mandante      24.843251
idade_media_titular_visitante     24.843251
valor_equipe_titular_mandante     24.819591
valor_equipe_titular_visitante    24.819591
gols_1_tempo_mandante             20.359636
gols_1_tempo_visitante          

In [18]:
df_clean[df_clean["chutes_mandante"].isnull()]["ano_campeonato"].value_counts().sort_index()

ano_campeonato
2003    456
2004    456
2005    418
2006    380
2007    380
2008    380
2009    380
2010    380
2011    380
2012    380
2013    380
2014    380
2015    380
2016    380
2017    380
2018    169
2019    212
2020      6
2021      5
2022      7
2023      2
2024     75
Name: count, dtype: int64

- As estatísticas detalhadas das partidas apresentam grande quantidade de valores ausentes entre 2003 e 2017.
- A partir de 2018, observa-se uma redução significativa na quantidade de dados ausentes.
- Os valores ausentes não foram substituídos por zero, pois a ausência de informação não significa necessariamente que o evento não ocorreu.

In [19]:
df_clean[df_clean["publico"].isnull()]["ano_campeonato"].value_counts().sort_index()

ano_campeonato
2003    456
2004    456
2005    418
2006    380
2016      1
Name: count, dtype: int64

- Os valores ausentes na coluna `publico` estão concentrados principalmente entre 2003 e 2006, indicando uma limitação histórica na disponibilidade dos dados.
- Apenas uma partida de 2016 apresentou ausência de informação de público.

In [20]:
missing_percent = (
    df_clean.isnull().sum() / len(df_clean) * 100
).sort_values(ascending=False)

missing_percent[missing_percent > 0]

chutes_fora_visitante             75.310541
defesas_visitante                 75.310541
defesas_mandante                  75.310541
faltas_mandante                   75.310541
faltas_visitante                  75.310541
chutes_bola_parada_visitante      75.310541
chutes_fora_mandante              75.310541
impedimentos_visitante            75.310541
impedimentos_mandante             75.310541
chutes_visitante                  75.310541
chutes_mandante                   75.310541
escanteios_mandante               75.310541
escanteios_visitante              75.310541
chutes_bola_parada_mandante       75.310541
publico_max                       50.538270
tecnico_mandante                  25.482077
tecnico_visitante                 25.482077
idade_media_titular_mandante      24.843251
idade_media_titular_visitante     24.843251
valor_equipe_titular_mandante     24.819591
valor_equipe_titular_visitante    24.819591
gols_1_tempo_mandante             20.359636
gols_1_tempo_visitante          

In [21]:
df_clean[
    df_clean["gols_mandante"].isnull() |
    df_clean["gols_visitante"].isnull()
]

,ano_campeonato,data,rodada,estadio,arbitro,publico,publico_max,time_mandante,time_visitante,tecnico_mandante,...,chutes_bola_parada_mandante,chutes_bola_parada_visitante,defesas_mandante,defesas_visitante,impedimentos_mandante,impedimentos_visitante,chutes_mandante,chutes_visitante,chutes_fora_mandante,chutes_fora_visitante
423,2016,2016-12-11,38,Arena Condá,Rodrigo D'Alonso Ferreira,NaN,22600.0,Chapecoense,Atlético-MG,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [22]:
df_clean = df_clean.dropna(
    subset=["gols_mandante", "gols_visitante"]
)

In [23]:
df_clean.shape

(8452, 35)

## Tratamento dos valores ausentes

- A base não apresentou linhas duplicadas.
- As estatísticas detalhadas das partidas apresentam grande quantidade de valores ausentes entre 2003 e 2017.
- Os valores ausentes das estatísticas não foram substituídos por zero, pois a ausência de informação não representa necessariamente a ocorrência de zero eventos.
- Foi identificada uma partida sem gols registrados nas colunas `gols_mandante` e `gols_visitante`.
- Como os gols são informações essenciais para as análises futuras, essa partida foi removida da base analítica.

In [46]:
df_clean.groupby("ano_campeonato")["chutes_mandante"].apply(
    lambda x: x.notnull().sum()
)

ano_campeonato
2003      0
2004      0
2005      0
2006      0
2007      0
2008      0
2009      0
2010      0
2011      0
2012      0
2013      0
2014      0
2015      0
2016      0
2017      0
2018    211
2019    168
2020    374
2021    375
2022    373
2023    378
2024    208
Name: chutes_mandante, dtype: int64

In [47]:
df_clean.groupby("ano_campeonato")["gols_mandante"].apply(
    lambda x: x.notnull().sum()
)

ano_campeonato
2003    456
2004    456
2005    418
2006    380
2007    380
2008    380
2009    380
2010    380
2011    380
2012    380
2013    380
2014    380
2015    380
2016    379
2017    380
2018    380
2019    380
2020    380
2021    380
2022    380
2023    380
2024    283
Name: gols_mandante, dtype: int64

In [48]:
pd.set_option("display.max_rows", None)

df_clean.isnull().sum().sort_values(ascending=False)

chutes_visitante                  6365
chutes_fora_mandante              6365
chutes_fora_visitante             6365
defesas_mandante                  6365
defesas_visitante                 6365
escanteios_visitante              6365
escanteios_mandante               6365
faltas_mandante                   6365
chutes_mandante                   6365
impedimentos_visitante            6365
impedimentos_mandante             6365
faltas_visitante                  6365
chutes_bola_parada_mandante       6365
chutes_bola_parada_visitante      6365
publico_max                       4272
tecnico_visitante                 2153
tecnico_mandante                  2153
idade_media_titular_visitante     2099
idade_media_titular_mandante      2099
valor_equipe_titular_mandante     2097
valor_equipe_titular_visitante    2097
gols_1_tempo_mandante             1720
gols_1_tempo_visitante            1720
arbitro                           1710
colocacao_mandante                1710
publico                  

In [49]:
#Novas colunas de data

df_clean["ano"] = df_clean["data"].dt.year
df_clean["mes"] = df_clean["data"].dt.month
df_clean["dia"] = df_clean["data"].dt.day

In [50]:
df_clean[["data", "ano", "mes", "dia"]].head()

,data,ano,mes,dia
0,2005-10-08,2005,10,8
1,2005-04-23,2005,4,23
2,2005-08-21,2005,8,21
3,2005-08-28,2005,8,28
4,2005-06-26,2005,6,26


In [51]:
# Gols mandante negativos
(df_clean["gols_mandante"] < 0).sum()

np.int64(0)

In [52]:

# Gols visitante negativos

(df_clean["gols_visitante"] < 0).sum()

np.int64(0)

In [53]:
#publico negativo

(df_clean["publico"] < 0).sum()

np.int64(0)

In [54]:
#rodada invalida

df_clean["rodada"].describe()

count    8452.000000
mean       19.384288
std        10.922928
min         1.000000
25%        10.000000
50%        19.000000
75%        29.000000
max        38.000000
Name: rodada, dtype: float64

In [55]:
#Quantos clubes existem

times = pd.concat([
    df_clean["time_mandante"],
    df_clean["time_visitante"]
]).unique()

len(times)

47

In [56]:
sorted(times)

['América-MG',
 'América-RN',
 'Athletico-PR',
 'Atlético-GO',
 'Atlético-MG',
 'Atlético-PR',
 'Avaí FC',
 'Barueri',
 'Botafogo',
 'Brasiliense-DF',
 'CSA',
 'Ceará SC',
 'Chapecoense',
 'Corinthians',
 'Coritiba FC',
 'Criciúma EC',
 'Cruzeiro',
 'Cuiabá-MT',
 'EC Bahia',
 'EC Vitória',
 'Figueirense FC',
 'Flamengo',
 'Fluminense',
 'Fortaleza',
 'Goiás',
 'Goiás EC',
 'Grêmio',
 'Guarani',
 'Internacional',
 'Ipatinga FC',
 'Joinville-SC',
 'Juventude',
 'Náutico',
 'Palmeiras',
 'Paraná',
 'Paysandu SC',
 'Ponte Preta',
 'Portuguesa',
 'RB Bragantino',
 'Santa Cruz',
 'Santo André',
 'Santos',
 'Santos FC',
 'Sport Recife',
 'São Caetano',
 'São Paulo',
 'Vasco da Gama']

In [57]:
#Remover espaços em branco no início e no final das strings

colunas_texto = [
    "time_mandante",
    "time_visitante",
    "tecnico_mandante",
    "tecnico_visitante",
    "estadio",
    "arbitro"
]

for coluna in colunas_texto:
    df_clean[coluna] = df_clean[coluna].str.strip()

In [58]:
# Coluna com o total de gols 

df_clean["total_gols"] = (
    df_clean["gols_mandante"] +
    df_clean["gols_visitante"]
)

In [59]:
df_clean[
    ["gols_mandante",
     "gols_visitante",
     "total_gols"]
].head()

,gols_mandante,gols_visitante,total_gols
0,6.0,1.0,7.0
1,0.0,2.0,2.0
2,0.0,0.0,0.0
3,0.0,4.0,4.0
4,0.0,1.0,1.0


In [60]:
def resultado(row):
    if row["gols_mandante"] > row["gols_visitante"]:
        return "Mandante"
    elif row["gols_mandante"] < row["gols_visitante"]:
        return "Visitante"
    else:
        return "Empate"

df_clean["resultado"] = df_clean.apply(resultado, axis=1)

In [61]:
df_clean["resultado"].value_counts()

resultado
Mandante     4182
Empate       2248
Visitante    2022
Name: count, dtype: int64

In [62]:
# Salvar a base de dados limpa em um novo arquivo CSV

df_clean.to_csv(
    "../data/processed/brasileirao_clean.csv",
    index=False
)

# Limpeza e Preparação dos Dados

## Decisões tomadas

- Os valores ausentes das estatísticas detalhadas foram mantidos, pois representam ausência de registro e não ausência do evento.
- Foram adicionadas as colunas `ano`, `mes`, `dia`, `total_gols` e `resultado`.
- A base tratada foi salva em `data/processed/brasileirao_clean.csv`.